src/librustc_lint/types.rs

```diff
@@ -1009,6 +1011,30 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ImproperCTypes {
             }
         }
     }
+
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'a, 'tcx>,
+        kind: hir::intravisit::FnKind<'tcx>,
+        decl: &'tcx hir::FnDecl,
+        _: &'tcx hir::Body,
+        _: Span,
+        hir_id: hir::HirId,
+    ) {
+        use hir::intravisit::FnKind;
+
+        let abi = match kind {
+            FnKind::ItemFn(_, _, header, ..) => (header.abi),
+            FnKind::Method(_, sig, ..) => (sig.header.abi),
+            _ => return,
+        };
+
+        // Check that the declaration of a fn is FFI-safe if `extern "C"`.
+        if abi == Abi::C {
```

Comment:
BTW we should probably also run the lint for other stable (non-rustc-specific) ABIs such as `win64`, `sysv`, `aapcs`, interrupt ABIs, etc., but this doesn't have to happen in this PR.

Reply:
https://github.com/rust-lang/rust/issues/65867

Reply:
I've addressed this in this PR.